### ISSUE TITLE
Protocol for Deprecation of Circuit Elements

### ISSUE THREAD

**Issue by satu0king** [CircuitVerse/CircuitVerse#1870]
There have been a few mistakes while creating some Circuit Elements. There have been some slight and some minor deviations in circuit behaviour. 

For example the splitter acts both like a splitter and a combiner. It seemed a good idea at the time, but we have had many complexities due to this. A separate Splitter and a Separate combiner seems the best way now. 

Some the mistakes are not correctable now as people are already using them and making changes will break their circuits. 

We need to find a way to either
1. Migrate elements from old->new
2. Just deprecate old ones and create new ones. Old circuits will continue to use old elements but new ones will use new elements. 

And other ideas?

**Comment by satu0king** [comment]
@gr455 @shubhankarsharma00 @abstrekt @mahmodHammad @drjyeh Any inputs here?

**Comment by shubhankarsharma00** [comment]
I had an idea we can have all modules have a version number AndGate-v1, NotGate-v1, RAM-v2 and save that while saving the project if it doesn't have a version number while loading we load element's version 1 else we load from new module's version.
Version clearance for old versions can be done in here:
https://github.com/CircuitVerse/CircuitVerse/blob/7b3ecb4e6141c681cebf0c1533176fc2bb7e2ce9/public/src/data/load.js#L22
We'll have to manage the sidebar to load all the elements of the latest version I guess, but apart from that, this seems pretty doable.

**Comment by satu0king** [comment]
@shubhankarsharma00 yes I was thinking on similar lines. But I think we need to discuss more on the details. Yes the sidebar can be configured to include only elements which don't have `.deprecated` as `true`. And we can configure formalType as 'AndGate-v2` and Human Type as `And Gate`. 

And about older types, we can use rectifyObject to *upgrade* some objects when possible. And in some cases *not to upgrade* and use old element itself. Basically, the ElementClass will be present but hidden and can be used by older circuits.

We should also account for maintainability, things are going to get much harder from here on, once we start doing this.

**Comment by tachyons** [comment]
We can also potentially think about parsing circuits in backend and update for existing circuits, but it can be tricky. 

We also have to think about circuitverse engine versioning with this feature
